### PR TITLE
Feature pd list #22

### PR DESCRIPTION
**Pull request by lkolacz**

### CODE REVIEW

File: django_api/django_api/apps/unicef/models.py [hunk 1]
```diff
@@ -1,19 +1,194 @@
...
+        'core.Location',
+        help_text='Select Administrative location for chosen level',
+    )
+    frequency = models.CharField(
+        max_length=3,
+        choices=FREQUENCY_LEVEL,
+        default=FREQUENCY_LEVEL.monthly,
+        verbose_name='Frequency of reporting'
+    )
+    reportable = models.ForeignKey(Reportable, null=True)
```
Comment (n1207n) [inline django_api/django_api/apps/unicef/models.py]: We have Reportable that can have LowerLevelOutput as a foreign key entity. This seems to be duplicate approach, but let's discuss with me and Sumit about this.

File: django_api/django_api/apps/unicef/views.py [hunk 2]
```diff
@@ -1,3 +1,55 @@
...
+
+
+        return queryset
+
+    def list(self, request, *args, **kwargs):
+        """
+        Get Programme Document list.
+        """
+        if request.GET:
+            form = ProgrammeDocumentForm(request.GET)
```
Comment (lkolacz) [inline django_api/django_api/apps/unicef/views.py]: done

File: django_api/django_api/apps/unicef/views.py [hunk 3]
```diff
@@ -1,3 +1,55 @@
...
+    ProgrammeDocumentSerializer,
+)
+
+from .models import ProgrammeDocument
+from .forms import ProgrammeDocumentForm
+
+
+class ProgrammeDocumentAPIView(ListAPIView):
+    """
+    Endpoint for getting Partner Details for overview tab.
```
Comment (lkolacz) [inline django_api/django_api/apps/unicef/views.py]: done

File: django_api/django_api/apps/unicef/models.py [hunk 4]
```diff
@@ -1,19 +1,194 @@
...
+    status = models.CharField(
+        max_length=256,
+        verbose_name='PD/SSFA status'
+    )
+    sections = models.ManyToManyField(Section)
+    contributing_to_cluser = models.BooleanField(
+        default=True,
+        verbose_name='Contributing to Cluser'
+    )
+    cluster_leds = models.ManyToManyField(
```
Comment (n1207n) [inline django_api/django_api/apps/unicef/models.py]: @chachra, what do you think about adding M2M field from Programme Document to Cluster? Also, there is another M2M field from Programme Document to locations as well.

I suppose we can access these from GenericForeignKey as well?

File: django_api/django_api/apps/unicef/models.py [hunk 5]
```diff
@@ -1,19 +1,194 @@
...
+
+        return self.__due_date
+
+    @property
+    def budget(self):
+        if self.__budget is not None:
+            return self.__budget
+        consumed = self.reportable.total
+        total = (
+            self.reportable.project and
```
Comment (n1207n) [inline django_api/django_api/apps/unicef/models.py]: This is a problematic with our GenericForeignkey setup. I don't think we can do double reversed related_name lookup, as this requires both Programme Document and PartnerProject links for IndicatorReport.

This one definitely needs a discussion. Let's talk about that tomorrow.

**Comment by n1207n** [comment]
Let's hold on merging this branch until we get to solve how to access budget info from PartnerProject for IndicatorReport that has link to Programme Document.

**Comment by rkrupinski** [comment]
@n1207n In terms of error handling:

tl;dr: we'll revisit this

I currently don't have any idea for a bulletproof architecture here. Handling it at redux level may be an option but that would require a centralised error communication mechanism (modal / global error msg / etc). Still, returning the original promise from the async action creator may come up handy since a component may want to add some custom failure logic. After moving the catch block to the action creator we'd then have to re-throw the original error (since the catch block swallows it) for the outer code to consume.

File: django_api/django_api/apps/indicator/models.py [hunk 6]
```diff
@@ -106,3 +106,8 @@ class IndicatorReport(TimeStampedModel):
         default=INDICATOR_REPORT_STATUS.ontrack,
         max_length=3
     )
+
+    programme_document = models.ForeignKey('unicef.ProgrammeDocument', null=True)
```
Comment (lkolacz) [inline django_api/django_api/apps/indicator/models.py]: I delete programme_document attr

File: django_api/django_api/apps/core/factories.py [hunk 7]
```diff
@@ -238,27 +241,43 @@ class Meta:
...
+    reportable = factory.SubFactory(ReportableToPartnerActivityFactory)
 
     class Meta:
         model = ProgrammeDocument
 
 
+class IndicatorReportFactory(factory.django.DjangoModelFactory):
+    title = factory.Sequence(lambda n: "indicator_report_%d" % n)
+    location = factory.SubFactory(LocationFactory)
+    reportable = factory.SubFactory(ReportableToPartnerActivityFactory)
```
Comment (lkolacz) [inline django_api/django_api/apps/core/factories.py]: @n1207n I tried a couple of times to make it happen .. but with `maximum recursion depth exceeded` problems .. so I guess it will be much easier when You merge Your branch to develop and then I will update it on here... currently it's waisting my time .. because it's not only a copy 2 factory classes as You mention ...

File: django_api/django_api/apps/core/factories.py [hunk 8]
```diff
@@ -238,27 +241,43 @@ class Meta:
...
     title = factory.Sequence(lambda n: "programme_document_%d" % n)
+    agreement = factory.Sequence(lambda n: "agreement_%d" % n)
+    reference_number = factory.Sequence(lambda n: "reference_number_%d" % n)
+    start_date = datetime.date.today()
+    end_date = datetime.date.today()+datetime.timedelta(days=70)
+    population_focus = factory.Sequence(lambda n: "population_focus%d" % n)
+    response_to_HRP = factory.Sequence(lambda n: "response_to_HRP%d" % n)
+    status = factory.Sequence(lambda n: "PD/SSFA status %d" % n)
+    frequency = FREQUENCY_LEVEL.weekly
+    reportable = factory.SubFactory(ReportableToPartnerActivityFactory)
```
Comment (lkolacz) [inline django_api/django_api/apps/core/factories.py]: @n1207n as I mentioned below

File: django_api/django_api/apps/unicef/models.py (hunk 1 above)
Comment (lkolacz) [inline django_api/django_api/apps/unicef/models.py]: @n1207n I don't remember the solution for this point .. should I delete it or it's fine ?

File: django_api/django_api/apps/unicef/models.py (hunk 5 above)
Comment (lkolacz) [inline django_api/django_api/apps/unicef/models.py]: @n1207n IMO this is the right way to get total money for partner.. and to getting consumed $ from `self.reportable.total`... If You see better (cheaper) relation please show me

File: django_api/django_api/apps/unicef/models.py (hunk 4 above)
Comment (chachra) [inline django_api/django_api/apps/unicef/models.py]: @n1207n I think so... lets try and stick to the data model we have, unless looking at it we are unable to query things easily I feel. Lmk. Adding redundant info might make data corrupt easily, but am fine with it if you strongly recommend it / there is no other way!

File: django_api/django_api/apps/core/factories.py [hunk 9]
```diff
@@ -197,6 +200,15 @@ class Meta:
         model = Reportable
 
 
+class ReportableToLowerLevelOutputFactory(ReportableFactory):
```
Comment (n1207n) [inline django_api/django_api/apps/core/factories.py]: This factory will create Reportable -> LowerLevelOutput -> CountryProgrammeOutput -> ProgrammeDocument all at once.

File: django_api/django_api/apps/core/factories.py [hunk 10]
```diff
@@ -238,27 +250,42 @@ class Meta:
...
+    frequency = FREQUENCY_LEVEL.weekly
 
     class Meta:
         model = ProgrammeDocument
 
 
+class IndicatorReportFactory(factory.django.DjangoModelFactory):
+    title = factory.Sequence(lambda n: "indicator_report_%d" % n)
+    location = factory.SubFactory(LocationFactory)
+    reportable = factory.SubFactory(ReportableToLowerLevelOutputFactory)
```
Comment (n1207n) [inline django_api/django_api/apps/core/factories.py]: This factory will create Reportable instance like I mentioned above when creating IndicatorReport object.

By calling this factory, we're creating IndicatorReport -> Reportable -> LowerLevelOutput -> CountryProgrammeOutput -> ProgrammeDocument.

All of the relevant objects are created and are linked properly.

File: django_api/django_api/apps/core/management/commands/_privates.py [hunk 11]
```diff
@@ -116,17 +120,19 @@ def generate_fake_data(quantity=3):
...
     print "{} ProgressReport objects created".format(quantity)
 
-    ProgrammeDocumentFactory.create_batch(quantity)
-    print "{} ProgrammeDocument objects created".format(quantity)
+    SectionFactory.create_batch(quantity)
+    print "{} Section objects created".format(quantity)
+
+    # TODO: more sens for IndicatorReport objects - important logic exist with frequency of PD
+    # IndicatorReport will create LowerLevelOutput, CountryProgrammeOutput, and ProgrammeDocument automatically
+    IndicatorReportFactory.create_batch(quantity)
```
Comment (n1207n) [inline django_api/django_api/apps/core/management/commands/_privates.py]: Note here that I deleted ProgrammeDocumentFactory calls here. Because IndicatorReportFactory will create ProgrammeDocument objects for me.

Therefore, your Section M2M field add is happening below by accessing indicator report's generic relation.

File: django_api/django_api/apps/partner/models.py [hunk 12]
```diff
@@ -151,15 +152,16 @@ class PartnerProject(TimeStampedModel):
     start_date = models.DateField()
     end_date = models.DateField()
     status = models.CharField(max_length=100)
-    budget = models.FloatField()
```
Comment (n1207n) [inline django_api/django_api/apps/partner/models.py]: Moving budget field to ProgrammeDocument model itself. As @chachra mentioned, there is no relation at all between PartnerProject and ProgrammeDocument.

In PMP world, budget is tied to ProgrammeDocument as a separate model. Therefore, we're mimicking same way in our PRP world.

File: django_api/django_api/apps/unicef/models.py [hunk 13]
```diff
@@ -1,19 +1,199 @@
...
+
+    @cached_property
+    def reportable_queryset(self):
+        return Reportable.objects.filter(
+            lower_level_outputs__indicator__programme_document=self)
+
+    @property
+    def reports_exists(self):
+        if self.__reports_exists is None:
+            self.__reports_exists = self.reportable_queryset.exists()
```
Comment (n1207n) [inline django_api/django_api/apps/unicef/models.py]: All of ProgrammeDocument python properties below access IndicatorReport instance.

Note how I am accessing latest indicator report via Reportable GenericRelation. This is the proper way to access IndicatorReport from ProgrammeDocument.

File: django_api/django_api/apps/unicef/models.py [hunk 14]
```diff
@@ -1,19 +1,199 @@
...
+            self.__due_date = due_report and due_report.time_period
+
+        return self.__due_date
+
+    @property
+    def calculated_budget(self):
+        if self.__budget is not None:
+            return self.__budget
+
+        total = self.budget
```
Comment (n1207n) [inline django_api/django_api/apps/unicef/models.py]: Now that budget field lives in ProgrammeDocument model itself, we no longer need PartnerProject links.

File: django_api/django_api/apps/unicef/views.py [hunk 15]
```diff
@@ -1,3 +1,28 @@
...
+from .filters import ProgrammeDocumentFilter
+
+
+class ProgrammeDocumentAPIView(ListAPIView):
+    """
+    Endpoint for getting Programme Document.
+    """
+    serializer_class = ProgrammeDocumentSerializer
+    permission_classes = (IsAuthenticated, )
+    filter_backends = (django_filters.rest_framework.DjangoFilterBackend,)
```
Comment (n1207n) [inline django_api/django_api/apps/unicef/views.py]: This is proper way of using Django-filter library, as described in both Django-REST-Framwork and django-filters. 

For now we're not doing anything dynamic to filter queryset. If we need to do that in the future, we can override filter_queryset method: http://www.django-rest-framework.org/api-guide/generic-views/#filter_querysetself-queryset

File: django_api/django_api/apps/core/factories.py (hunk 9 above)
Comment (lkolacz) [inline django_api/django_api/apps/core/factories.py]: great

File: django_api/django_api/apps/core/management/commands/_privates.py (hunk 11 above)
Comment (lkolacz) [inline django_api/django_api/apps/core/management/commands/_privates.py]: Thx Silin for great description and sharing knowledge with me :)
I'm appreciate 👍

File: django_api/django_api/apps/unicef/models.py (hunk 14 above)
Comment (lkolacz) [inline django_api/django_api/apps/unicef/models.py]: sure thing :)

**Comment by lkolacz** [comment]
@n1207n I test it and bring improvements for filed budget (float->decimal) .. improve factory budget was in PartnerProgram when he should be in ProgrammeDocument as You mention in code .. 
I also write a cover unit test :)
I guess I full fill our comments
Thx again 👍

**Comment by n1207n** [comment]
I know this PR is already merged, but this PR got merged with failed CI and no polymer test cases.

From now on, I would like to see the branch passes CI testing and test cases for backend and frontend side.
